[F] Add searching and filtering for project resources

SMaxOwok opened the pull request:
[Delivers #139241941]

zdavis commented on client/src/components/frontend/ResourceList/Filters.js:
```diff
@@ -1,17 +1,62 @@
+    filterChangeHandler: PropTypes.func.isRequired
   };
 
-  constructor() {
```
You've changed this component so that it needs `kinds` and `tags` props to function properly, but you haven't gone and made the change to the parent that uses this component to ensure that it still functions. Take a look at `components/frontend/ResourceCollection/Detail`. That component includes `ResourceList.Filters`, but doesn't pass it tags — so when you drill into a collection on a project, you end up with the map in `ResourceList.Filters` render method throwing an exception.

Think about how you can make the filters component more flexible and error-resistant. If it's not passed tags or kinds, shouldn't it just not render those parts of the filters? Or, if it's going to always render them, then at least set default props on it:

```
static defaultProps = {
  kind: [],
  tags: []
};
```

zdavis commented on api/app/controllers/api/v1/collections/relationships/resources_controller.rb:
```diff
@@ -8,8 +8,9 @@ class ResourcesController < ApplicationController
-                       .query(with_pagination!(resource_filter_params))
+            filter_params =
+              with_pagination!(resource_filter_params).merge(collection: @collection.id)
+            Resource.filter(with_pagination!(filter_params))
```
Look at the changes I made to how filtering works. With those changes, I think it should be possible to change this call to look like this:

`Resource.filter(with_pagination!(filter_params), @collection.resources)`

I think this is a cleaner solution, as we can use the association scope rather than fudging the params. Try it and confirm that it works.

zdavis commented on api/app/models/concerns/filterable.rb:
```diff
@@ -3,14 +3,63 @@ module Filterable
 
```
Look at my extended commit message for my thinking on this. I like where you're going with the filtering—getting the IDs from the active record query and passing them to elastic search if necessary—but having that logic spread out across all our models is going to get kind of crazy. I've refactored the filterable concern so that all filtering (and pagination) logic is in one place. The only thing that stays on the models are the named scopes, which seems correct. This is still a little complex, and we may want to tweak it, but at least it's now DRY.

zdavis commented on api/app/models/concerns/paginated.rb:
```diff
@@ -1,25 +0,0 @@
-# Model concern that includes pagination scope
```
I moved all this into `Filterable`.

zdavis commented on api/app/models/project.rb:
```diff
@@ -135,6 +117,12 @@ def resource_kinds
       .to_a.pluck(:kind)
   end
 
+  def resource_tags
```
Hmmm.... I don't love this approach, as it introduces what could potentially be a hefty query for every single project serialization. It's fine for now, but I wonder if we shouldn't instead do a separate request for tags by project when we need that.

zdavis commented on api/app/models/resource.rb:
```diff
@@ -63,6 +63,22 @@ class Resource < ApplicationRecord
+    return all unless collection.present?
+    where(collection: collection)
+  }
+  scope :with_order, lambda { |by|
```
We may as well handle order as a named scope, just like we do with the where clauses.

zdavis commented on api/spec/models/resource_spec.rb:
```diff
@@ -25,4 +25,24 @@
+      results = Resource.filter({project: @project_b})
+      expect(results.length).to be 1
+    end
+
```
I just added a super basic test... would be nice to have a few more.

zdavis commented on client/src/components/frontend/Project/Resources.js:
```diff
@@ -3,12 +3,13 @@ import { ResourceList } from 'components/frontend';
     resources: PropTypes.array,
-    pagination: PropTypes.object
+    pagination: PropTypes.object,
+    filterChange: PropTypes.func.isRequired
```
Double check to be sure that everything currently using this component is passing this new required prop in.

zdavis commented on client/src/components/frontend/ResourceList/Filters.js:
```diff
@@ -1,17 +1,62 @@
+  }
+
+  setFilters(event, label) {
+    const key = event.target.value;
```
Seems confusing to refer to value as key here.

zdavis commented on client/src/components/frontend/ResourceList/Filters.js:
```diff
@@ -1,17 +1,62 @@
+    const key = event.target.value;
+    const filter = Object.assign({}, this.state.filter);
+    const inputs = Object.assign({}, this.state.inputs);
+    if (key === "") {
```
There's something about this method I'm not crazy about... it feels verbose, for one. Also, there's the fact that this will always update state, even if nothing has changed in the inputs or the filters. I'm not saying rewrite it, I just think it could probably be tightened up a bit.

zdavis commented on client/src/components/frontend/ResourceList/Filters.js:
```diff
@@ -41,36 +93,40 @@ export default class ResourceListFilters extends Component {
-              <option>Japan</option>
-              <option>Photography</option>
-              <option>Second Interview</option>
+              {this.props.tags.map((tag, index) => {
```
If `tags` does not exist, this will choke.

SMaxOwok commented:
I added a couple more tests and refactored the setFilters method to be a case switch statement and only update when there is a value/label.  I'm not totally certain about your comment about always updating state, since it only fires when a filter is changed.  Would you mind clarifying?

SMaxOwok commented on client/src/components/frontend/Resource/Card.js:
```diff
@@ -74,12 +75,21 @@ export default class ResourceCard extends Component {
-      const crid = this.props.resource.id;
       const pid = context.relationships.project.id;
       const cid = context.id;
+      const crid = this.getCollectionResourceId(this.props.resource);
```
I had to include the CollectionResources for a Resource because of the different IDs.  Is there a better way we could handle this or does this seem appropriate? It seems like either way we get the resources (by CollectionResource or Resources) we need to reference the association.

zdavis replied:
That seems OK.

zdavis commented:
Pressing return in the search input submits the form and reloads the page. You need to capture the form onSubmit event.
![screen shot 2017-02-24 at 1 02 32 pm](https://cloud.githubusercontent.com/assets/255909/23320892/8994796c-fa91-11e6-892c-b1a8fcd551d9.png)

zdavis commented:
In general the search input needs additional testing and review. Give it another thorough pass.

- Changing sort has no impact on the collection view.
- The keyword search shouldn't be typeahead. It needs to require a submit to update results. 
- If you type the letter "a" (or probably any single letter) into the keyword input, you can't delete it.

SMaxOwok commented:
Addressed those issues and fixed the missing collection resource default sorting order.
